astroid/nodes/node_classes.py

```diff
@@ -4728,6 +4756,30 @@ def postinit(self, values: list[NodeNG] | None = None) -> None:
     def get_children(self):
         yield from self.values
 
+    def _infer(
+        self, context: InferenceContext | None = None, **kwargs: Any
+    ) -> Generator[InferenceResult, None, InferenceErrorInfo | None]:
+        yield from self._infer_from_values(self.values)
+
+    @classmethod
+    def _infer_from_values(
+        cls, nodes: list[NodeNG], context: InferenceContext | None = None, **kwargs: Any
+    ) -> Generator[InferenceResult, None, InferenceErrorInfo | None]:
+        if len(nodes) == 1:
+            yield from nodes[0]._infer(context, **kwargs)
+            return
+        for prefix in nodes[0]._infer(context, **kwargs):
+            for suffix in cls._infer_from_values(nodes[1:], context, **kwargs):
+                result = ""
+                for node in (prefix, suffix):
+                    if node is util.Uninferable:
+                        result += "{Uninferable}"
```

Comment:
It feels like we should exit early here and not continue to product "half" matching values

Reply:
That's philosophical, depending on the strategy for dealing with partially inferred collections of values

Reply:
It would help to make a decision about this if we could see a test case. At the moment, I do not think we should manufacture a constant that would never exist at runtime with the substring "Uninferable" in it.

Reply:
There's "coincidental" coverage of this line in `test_typing_extensions_types` based on the current implementation of typing, but it would be better to have explicit coverage.

Reply:
I agree with Jacob. I think `Uninferable` makes sense to those that have worked on `astroid` and `pylint`, but this will also be shown to users who might have no idea what it means. I would be in favour of exiting early and not creating partially correct values.

Reply:
Done

Reply:
It is useful to know which fragment(s) of a JoinedStr cause inference failure, but I agree that i's misleading to return half-baked strings.